### PR TITLE
Fix/improve available responsibles and assignees performance

### PR DESCRIPTION
**Pull request by ulferts**
Turns the N+1 queries here:

```
Started GET "/api/v3/projects/1/available_assignees" for 127.0.0.1 at 2015-09-23 13:29:42 +0200
  User Load (0.3ms)  SELECT  "users".* FROM "users" WHERE "users"."type" IN ('User', 'AnonymousUser', 'DeletedUser') AND "users"."id" = $1 LIMIT 1  [["id", 1]]
   (0.3ms)  SELECT MAX("settings"."updated_on") FROM "settings"
  Project Load (0.2ms)  SELECT  "projects".* FROM "projects" WHERE "projects"."identifier" = $1  ORDER BY "projects"."id" ASC LIMIT 1  [["identifier", "1"]]
  Project Load (0.2ms)  SELECT  "projects".* FROM "projects" WHERE "projects"."id" = $1 LIMIT 1  [["id", 1]]
   (0.1ms)  SELECT "enabled_modules"."name" FROM "enabled_modules" WHERE "enabled_modules"."project_id" = $1  [["project_id", 1]]
  SQL (0.3ms)  SELECT "members"."id" AS t0_r0, "members"."user_id" AS t0_r1, "members"."project_id" AS t0_r2, "members"."created_on" AS t0_r3, "members"."mail_notification" AS t0_r4, "users"."id" AS t1_r0, "users"."login" AS t1_r1, "users"."firstname" AS t1_r2, "users"."lastname" AS t1_r3, "users"."mail" AS t1_r4, "users"."admin" AS t1_r5, "users"."status" AS t1_r6, "users"."last_login_on" AS t1_r7, "users"."language" AS t1_r8, "users"."auth_source_id" AS t1_r9, "users"."created_on" AS t1_r10, "users"."updated_on" AS t1_r11, "users"."type" AS t1_r12, "users"."identity_url" AS t1_r13, "users"."mail_notification" AS t1_r14, "users"."first_login" AS t1_r15, "users"."force_password_change" AS t1_r16, "users"."failed_login_count" AS t1_r17, "users"."last_failed_login_on" AS t1_r18, "roles"."id" AS t2_r0, "roles"."name" AS t2_r1, "roles"."position" AS t2_r2, "roles"."assignable" AS t2_r3, "roles"."builtin" AS t2_r4, "roles"."permissions" AS t2_r5 FROM "members" LEFT OUTER JOIN "users" ON "users"."id" = "members"."user_id" LEFT OUTER JOIN "member_roles" ON "member_roles"."member_id" = "members"."id" LEFT OUTER JOIN "roles" ON "roles"."id" = "member_roles"."role_id" WHERE "members"."project_id" = $1 AND ((users.type='User') AND users.status=1 AND roles.assignable = 't')  [["project_id", 1]]
  UserPreference Load (0.2ms)  SELECT  "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT 1  [["user_id", 7]]
  UserPreference Load (0.2ms)  SELECT  "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT 1  [["user_id", 4]]
  UserPreference Load (0.2ms)  SELECT  "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT 1  [["user_id", 1]]
  UserPreference Load (0.2ms)  SELECT  "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT 1  [["user_id", 2]]
  UserPreference Load (0.2ms)  SELECT  "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT 1  [["user_id", 6]]
  UserPreference Load (0.2ms)  SELECT  "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT 1  [["user_id", 5]]


Started GET "/api/v3/projects/1/available_responsibles" for 127.0.0.1 at 2015-09-23 13:29:43 +0200
  User Load (0.3ms)  SELECT  "users".* FROM "users" WHERE "users"."type" IN ('User', 'AnonymousUser', 'DeletedUser') AND "users"."id" = $1 LIMIT 1  [["id", 1]]
   (0.5ms)  SELECT MAX("settings"."updated_on") FROM "settings"
  Project Load (0.3ms)  SELECT  "projects".* FROM "projects" WHERE "projects"."identifier" = $1  ORDER BY "projects"."id" ASC LIMIT 1  [["identifier", "1"]]
  Project Load (0.2ms)  SELECT  "projects".* FROM "projects" WHERE "projects"."id" = $1 LIMIT 1  [["id", 1]]
   (0.2ms)  SELECT "enabled_modules"."name" FROM "enabled_modules" WHERE "enabled_modules"."project_id" = $1  [["project_id", 1]]
  SQL (0.4ms)  SELECT "members"."id" AS t0_r0, "members"."user_id" AS t0_r1, "members"."project_id" AS t0_r2, "members"."created_on" AS t0_r3, "members"."mail_notification" AS t0_r4, "users"."id" AS t1_r0, "users"."login" AS t1_r1, "users"."firstname" AS t1_r2, "users"."lastname" AS t1_r3, "users"."mail" AS t1_r4, "users"."admin" AS t1_r5, "users"."status" AS t1_r6, "users"."last_login_on" AS t1_r7, "users"."language" AS t1_r8, "users"."auth_source_id" AS t1_r9, "users"."created_on" AS t1_r10, "users"."updated_on" AS t1_r11, "users"."type" AS t1_r12, "users"."identity_url" AS t1_r13, "users"."mail_notification" AS t1_r14, "users"."first_login" AS t1_r15, "users"."force_password_change" AS t1_r16, "users"."failed_login_count" AS t1_r17, "users"."last_failed_login_on" AS t1_r18, "roles"."id" AS t2_r0, "roles"."name" AS t2_r1, "roles"."position" AS t2_r2, "roles"."assignable" AS t2_r3, "roles"."builtin" AS t2_r4, "roles"."permissions" AS t2_r5 FROM "members" LEFT OUTER JOIN "users" ON "users"."id" = "members"."user_id" LEFT OUTER JOIN "member_roles" ON "member_roles"."member_id" = "members"."id" LEFT OUTER JOIN "roles" ON "roles"."id" = "member_roles"."role_id" WHERE "members"."project_id" = $1 AND ((users.type='User') AND users.status=1 AND roles.assignable = 't')  [["project_id", 1]]
  UserPreference Load (0.2ms)  SELECT  "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT 1  [["user_id", 7]]
  UserPreference Load (0.3ms)  SELECT  "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT 1  [["user_id", 4]]
  UserPreference Load (0.3ms)  SELECT  "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT 1  [["user_id", 1]]
  UserPreference Load (0.2ms)  SELECT  "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT 1  [["user_id", 2]]
  UserPreference Load (0.2ms)  SELECT  "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT 1  [["user_id", 6]]
  UserPreference Load (0.2ms)  SELECT  "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT 1  [["user_id", 5]]
```

Into this:

```
Started GET "/api/v3/projects/1/available_assignees" for 127.0.0.1 at 2015-09-23 13:27:18 +0200
  User Load (0.3ms)  SELECT  "users".* FROM "users" WHERE "users"."type" IN ('User', 'AnonymousUser', 'DeletedUser') AND "users"."id" = $1 LIMIT 1  [["id", 1]]
   (0.3ms)  SELECT MAX("settings"."updated_on") FROM "settings"
  Project Load (0.2ms)  SELECT  "projects".* FROM "projects" WHERE "projects"."identifier" = $1  ORDER BY "projects"."id" ASC LIMIT 1  [["identifier", "1"]]
  Project Load (0.2ms)  SELECT  "projects".* FROM "projects" WHERE "projects"."id" = $1 LIMIT 1  [["id", 1]]
   (0.2ms)  SELECT "enabled_modules"."name" FROM "enabled_modules" WHERE "enabled_modules"."project_id" = $1  [["project_id", 1]]
   (1.2ms)  SELECT COUNT(DISTINCT "users"."id") FROM "users" LEFT OUTER JOIN "members" "members_users" ON "members_users"."user_id" = "users"."id" LEFT OUTER JOIN "member_roles" ON "member_roles"."member_id" = "members_users"."id" LEFT OUTER JOIN "roles" ON "roles"."id" = "member_roles"."role_id" LEFT OUTER JOIN "user_preferences" ON "user_preferences"."user_id" = "users"."id" INNER JOIN "members" ON "users"."id" = "members"."user_id" WHERE "members"."project_id" = $1 AND ((users.type='User') AND users.status=1 AND roles.assignable = 't')  [["project_id", 1]]
  CACHE (0.1ms)  SELECT COUNT(DISTINCT "users"."id") FROM "users" LEFT OUTER JOIN "members" "members_users" ON "members_users"."user_id" = "users"."id" LEFT OUTER JOIN "member_roles" ON "member_roles"."member_id" = "members_users"."id" LEFT OUTER JOIN "roles" ON "roles"."id" = "member_roles"."role_id" LEFT OUTER JOIN "user_preferences" ON "user_preferences"."user_id" = "users"."id" INNER JOIN "members" ON "users"."id" = "members"."user_id" WHERE "members"."project_id" = $1 AND ((users.type='User') AND users.status=1 AND roles.assignable = 't')  [["project_id", 1]]
  SQL (1.5ms)  SELECT "users"."id" AS t0_r0, "users"."login" AS t0_r1, "users"."firstname" AS t0_r2, "users"."lastname" AS t0_r3, "users"."mail" AS t0_r4, "users"."admin" AS t0_r5, "users"."status" AS t0_r6, "users"."last_login_on" AS t0_r7, "users"."language" AS t0_r8, "users"."auth_source_id" AS t0_r9, "users"."created_on" AS t0_r10, "users"."updated_on" AS t0_r11, "users"."type" AS t0_r12, "users"."identity_url" AS t0_r13, "users"."mail_notification" AS t0_r14, "users"."first_login" AS t0_r15, "users"."force_password_change" AS t0_r16, "users"."failed_login_count" AS t0_r17, "users"."last_failed_login_on" AS t0_r18, "members_users"."id" AS t1_r0, "members_users"."user_id" AS t1_r1, "members_users"."project_id" AS t1_r2, "members_users"."created_on" AS t1_r3, "members_users"."mail_notification" AS t1_r4, "roles"."id" AS t2_r0, "roles"."name" AS t2_r1, "roles"."position" AS t2_r2, "roles"."assignable" AS t2_r3, "roles"."builtin" AS t2_r4, "roles"."permissions" AS t2_r5, "user_preferences"."id" AS t3_r0, "user_preferences"."user_id" AS t3_r1, "user_preferences"."others" AS t3_r2, "user_preferences"."hide_mail" AS t3_r3, "user_preferences"."time_zone" AS t3_r4, "user_preferences"."impaired" AS t3_r5 FROM "users" LEFT OUTER JOIN "members" "members_users" ON "members_users"."user_id" = "users"."id" LEFT OUTER JOIN "member_roles" ON "member_roles"."member_id" = "members_users"."id" LEFT OUTER JOIN "roles" ON "roles"."id" = "member_roles"."role_id" LEFT OUTER JOIN "user_preferences" ON "user_preferences"."user_id" = "users"."id" INNER JOIN "members" ON "users"."id" = "members"."user_id" WHERE "members"."project_id" = $1 AND ((users.type='User') AND users.status=1 AND roles.assignable = 't')  ORDER BY firstname, lastname  [["project_id", 1]]


Started GET "/api/v3/projects/1/available_responsibles" for 127.0.0.1 at 2015-09-23 13:27:18 +0200
  User Load (0.3ms)  SELECT  "users".* FROM "users" WHERE "users"."type" IN ('User', 'AnonymousUser', 'DeletedUser') AND "users"."id" = $1 LIMIT 1  [["id", 1]]
   (0.3ms)  SELECT MAX("settings"."updated_on") FROM "settings"
  Project Load (0.3ms)  SELECT  "projects".* FROM "projects" WHERE "projects"."identifier" = $1  ORDER BY "projects"."id" ASC LIMIT 1  [["identifier", "1"]]
  Project Load (0.2ms)  SELECT  "projects".* FROM "projects" WHERE "projects"."id" = $1 LIMIT 1  [["id", 1]]
   (0.2ms)  SELECT "enabled_modules"."name" FROM "enabled_modules" WHERE "enabled_modules"."project_id" = $1  [["project_id", 1]]
   (1.1ms)  SELECT COUNT(DISTINCT "users"."id") FROM "users" LEFT OUTER JOIN "members" "members_users" ON "members_users"."user_id" = "users"."id" LEFT OUTER JOIN "member_roles" ON "member_roles"."member_id" = "members_users"."id" LEFT OUTER JOIN "roles" ON "roles"."id" = "member_roles"."role_id" LEFT OUTER JOIN "user_preferences" ON "user_preferences"."user_id" = "users"."id" INNER JOIN "members" ON "users"."id" = "members"."user_id" WHERE "members"."project_id" = $1 AND ((users.type='User') AND users.status=1 AND roles.assignable = 't')  [["project_id", 1]]
  CACHE (0.0ms)  SELECT COUNT(DISTINCT "users"."id") FROM "users" LEFT OUTER JOIN "members" "members_users" ON "members_users"."user_id" = "users"."id" LEFT OUTER JOIN "member_roles" ON "member_roles"."member_id" = "members_users"."id" LEFT OUTER JOIN "roles" ON "roles"."id" = "member_roles"."role_id" LEFT OUTER JOIN "user_preferences" ON "user_preferences"."user_id" = "users"."id" INNER JOIN "members" ON "users"."id" = "members"."user_id" WHERE "members"."project_id" = $1 AND ((users.type='User') AND users.status=1 AND roles.assignable = 't')  [["project_id", 1]]
  SQL (1.4ms)  SELECT "users"."id" AS t0_r0, "users"."login" AS t0_r1, "users"."firstname" AS t0_r2, "users"."lastname" AS t0_r3, "users"."mail" AS t0_r4, "users"."admin" AS t0_r5, "users"."status" AS t0_r6, "users"."last_login_on" AS t0_r7, "users"."language" AS t0_r8, "users"."auth_source_id" AS t0_r9, "users"."created_on" AS t0_r10, "users"."updated_on" AS t0_r11, "users"."type" AS t0_r12, "users"."identity_url" AS t0_r13, "users"."mail_notification" AS t0_r14, "users"."first_login" AS t0_r15, "users"."force_password_change" AS t0_r16, "users"."failed_login_count" AS t0_r17, "users"."last_failed_login_on" AS t0_r18, "members_users"."id" AS t1_r0, "members_users"."user_id" AS t1_r1, "members_users"."project_id" AS t1_r2, "members_users"."created_on" AS t1_r3, "members_users"."mail_notification" AS t1_r4, "roles"."id" AS t2_r0, "roles"."name" AS t2_r1, "roles"."position" AS t2_r2, "roles"."assignable" AS t2_r3, "roles"."builtin" AS t2_r4, "roles"."permissions" AS t2_r5, "user_preferences"."id" AS t3_r0, "user_preferences"."user_id" AS t3_r1, "user_preferences"."others" AS t3_r2, "user_preferences"."hide_mail" AS t3_r3, "user_preferences"."time_zone" AS t3_r4, "user_preferences"."impaired" AS t3_r5 FROM "users" LEFT OUTER JOIN "members" "members_users" ON "members_users"."user_id" = "users"."id" LEFT OUTER JOIN "member_roles" ON "member_roles"."member_id" = "members_users"."id" LEFT OUTER JOIN "roles" ON "roles"."id" = "member_roles"."role_id" LEFT OUTER JOIN "user_preferences" ON "user_preferences"."user_id" = "users"."id" INNER JOIN "members" ON "users"."id" = "members"."user_id" WHERE "members"."project_id" = $1 AND ((users.type='User') AND users.status=1 AND roles.assignable = 't')  ORDER BY firstname, lastname  [["project_id", 1]]
```

It does this by eager loading the user preferences. 
A hack is included in the PR, though. For the eager loading, the `:preference` association has to be on the Principal model. But the UserPreference is currently very User specific. Therefore, all methods added by having a `has_one` association are undefined in the Group model. 

Additionally, it adds a `order_by_name` scope on the Principal model which will return the principals ordered by the current setting for displaying names. This should speed up sorting as it can then be done in the DB.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://ulferts-fix-improve-available-responsibles-and-assignees-perfor.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://ulferts-fix-improve-available-responsibles-and-assignees-perfor.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: app/models/principal.rb [hunk 1]
```diff
@@ -114,6 +118,10 @@ def self.visible_by_condition(principal)
           project_ids)
   end
 
+  def self.order_by_name
+    order(User::USER_FORMATS_STRUCTURE[Setting.user_format].map(&:to_s))
+  end
```
Comment (NobodysNightmare) [inline app/models/principal.rb:123]: why isn't that a scope? Probably `ordered_by_name`, for nicer reading experience...

**Comment by NobodysNightmare** [comment]
LGTM, albeit the hack :horse: 

Waiting for travis.

**Comment by NobodysNightmare** [comment]
@ulferts Looks like there is a bunch of legitimate failures :(

**Comment by ulferts** [comment]
Thanks for the _ping_ @NobodysNightmare. Most seem to be caused by mocks not being adapted. Will take a look.

File: spec/views/api/experimental/projects/show_api_json_spec.rb [hunk 2]
```diff
@@ -29,16 +29,15 @@
...
+  let(:principals) { FactoryGirl.build_list(:principal, 3) }
   let(:types)     { FactoryGirl.build_list(:type,   2) }
 
   let(:project)   {
-    FactoryGirl.build(:project,
-                      possible_responsible_members: members,
-                      possible_assignee_members:    members,
-                      types:                        types
-                     )
+    project = FactoryGirl.build(:project,
```
Comment (NobodysNightmare) [inline spec/views/api/experimental/projects/show_api_json_spec.rb]: What do you need that local for?

**Comment by NobodysNightmare** [comment]
good job on removing that spec duplication while preserving searchability.

Will merge when travis goes green. :horse:

**Pull request merged**